### PR TITLE
Add `CosIdKeyGenerateAlgorithm` / `CosIdIntervalShardingAlgorithm` /`CosIdModShardingAlgorithm`

### LINKED ISSUE #14047: IntervalShardingAlgorithm performance is too bad
#### When I was about to use `IntervalShardingAlgorithm` to integrate [CosId](https://github.com/Ahoo-Wang/CosId), I checked the source code and found the following problems:

- Ease of use: The `IntervalShardingAlgorithm` implementation is to first convert to a string and then convert to `LocalDateTime`, the conversion success rate is affected by the time formatting characters
- Performance: The method of nested traversal checking determines whether the conditions are met, and it is accompanied by the conversion process of `LocalDateTime`. The performance problem is fatal. The performance of `PreciseShardingValue` is even lower than `7000 ops/s` (it is lower than the storage layer `MySql`, *ShardingSphere-JDBC* becomes the bottleneck, which is obviously unbearable)

[Code implementation of benchmark report](https://github.com/Ahoo-Wang/CosId/tree/main/cosid-shardingsphere/src/jmh)

> I even doubt whether there is a problem with the way the benchmark report is implemented,But I have tried my best to eliminate the test noise

Interval-based-sharding-algorithm-jmh

``` shell
gradle cosid-shardingsphere:jmh
```
```
# JMH version: 1.29
# VM version: JDK 11.0.13, OpenJDK 64-Bit Server VM, 11.0.13+8-LTS
# VM options: -Dfile.encoding=UTF-8 -Djava.io.tmpdir=/work/CosId/cosid-shardingsphere/build/tmp/jmh -Duser.country=CN -Duser.language=zh -Duser.variant
# Blackhole mode: full + dont-inline hint
# Warmup: 1 iterations, 10 s each
# Measurement: 1 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
Benchmark                                                          Mode  Cnt         Score   Error  Units
IntervalShardingAlgorithmBenchmark.cosid_precise_local_date_time  thrpt       66276995.822          ops/s
IntervalShardingAlgorithmBenchmark.cosid_precise_timestamp        thrpt       24841952.001          ops/s
IntervalShardingAlgorithmBenchmark.cosid_range_local_date_time    thrpt        3344013.803          ops/s
IntervalShardingAlgorithmBenchmark.cosid_range_timestamp          thrpt        2846453.298          ops/s
IntervalShardingAlgorithmBenchmark.office_precise_timestamp       thrpt           6286.861          ops/s
IntervalShardingAlgorithmBenchmark.office_range_timestamp         thrpt           2302.986          ops/s
```

#### So I re-implemented the time range sharding algorithm based on time interval to improve the ease of use and performance.
[https://github.com/Ahoo-Wang/CosId/releases/tag/v1.4.5](https://github.com/Ahoo-Wang/CosId/releases/tag/v1.4.5)

CosIdIntervalShardingAlgorithm

- DateIntervalShardingAlgorithm
  - type: COSID_INTERVAL_DATE
- LocalDateTimeIntervalShardingAlgorithm
  - type: COSID_INTERVAL_LDT
- TimestampIntervalShardingAlgorithm
  - type: COSID_INTERVAL_TS
- TimestampOfSecondIntervalShardingAlgorithm
  - type: COSID_INTERVAL_TS_SECOND
- SnowflakeIntervalShardingAlgorithm
  - type: COSID_INTERVAL_SNOWFLAKE

### Is your feature request related to a problem?

NO

### Describe the feature you would like.

If you think this implementation is good, I can submit a PR

### PR DESCRIPTION
**Pull request by Ahoo-Wang**
Fixes #14047 .

Changes proposed in this pull request:
- add `CosIdKeyGenerateAlgorithm`
- add `CosIdIntervalShardingAlgorithm`
- add `CosIdModShardingAlgorithm`

https://github.com/Ahoo-Wang/CosId#shardingsphere-plugin

---
This pull request does not include (I will finish the rest of the issue in the next time.):
> `update original SNOWFLAKE`

> Integrate COSID_SEGMENT and COSID_SEGMENT_CHAIN with ShardingSphere deeply,just reuse reg-center of cluster mode.

### CODE REVIEW

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+public final class CosIdKeyGenerateAlgorithm implements KeyGenerateAlgorithm {
+
+    public static final String TYPE = CosId.COSID.toUpperCase();
+
+    private Properties props = new Properties();
+
+    private volatile LazyIdGenerator cosIdProvider;
+
+    /**
+     * Get properties.
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java:38]: Unnecessary comments for override methods, please remove all comments

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.Collection;
+import java.util.Date;
+import java.util.Properties;
+
+/**
+ * Interval-based time range sharding algorithm.
+ */
+public final class CosIdIntervalShardingAlgorithm implements StandardShardingAlgorithm<Comparable<?>> {
+
+    public static final String TYPE = CosIdAlgorithm.TYPE_PREFIX + "INTERVAL";
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java:42]: Please confirm whether all variables need `public`

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    public static final String INTERVAL_UNIT_KEY = "datetime-interval-unit";
+
+    public static final String INTERVAL_AMOUNT_KEY = "datetime-interval-amount";
+
+    public static final String TIMESTAMP_SECOND_TYPE = "SECOND";
+
+    public static final String ZONE_ID_KEY = "zone-id";
+
+    /**
+     * type of timestamp.
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java:65]: Unnecessary comment

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        return props;
+    }
+
+    @Override
+    public void setProps(final Properties props) {
+        this.props = props;
+    }
+
+    /**
+     * Initialize algorithm.
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java:84]: same as above

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    public Collection<String> doSharding(final Collection<String> availableTargetNames, final RangeShardingValue<Comparable<?>> shardingValue) {
+        Range<LocalDateTime> shardingRangeTime = convertRangeShardingValue(shardingValue.getValueRange());
+        return algorithmConfig.getIntervalTimeline().sharding(shardingRangeTime);
+    }
+
+    private LocalDateTime convertShardingValue(final Comparable<?> shardingValue) {
+        if (shardingValue instanceof LocalDateTime) {
+            return (LocalDateTime) shardingValue;
+        }
+
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java:136]: please remove unnecessary empty line

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithm.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    @Override
+    public void setProps(final Properties props) {
+        this.props = props;
+    }
+
+    /**
+     * Sharding.
+     *
+     * @param availableTargetNames available data sources or table names
+     * @param shardingValue        sharding value
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithm.java:73]: same as above.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithmTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import me.ahoo.cosid.segment.IdSegmentDistributor;
+import org.apache.shardingsphere.sharding.algorithm.sharding.cosid.CosIdAlgorithm;
+import org.junit.Test;
+
+import java.util.Properties;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public class CosIdKeyGenerateAlgorithmTest {
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithmTest.java:30]: class should be final.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithmTest.java [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+    static Iterable<Object[]> rangeArgsProviderAsDate() {
+        return rangeArgsProvider(ldt -> new Date(ldt.toInstant(ZONE_OFFSET_SHANGHAI).toEpochMilli()));
+    }
+
+    static Iterable<Object[]> rangeArgsProviderAsTimestamp() {
+        return rangeArgsProvider(ldt -> ldt.toInstant(ZONE_OFFSET_SHANGHAI).toEpochMilli());
+    }
+
+    @RunWith(Parameterized.class)
+    public static class LocalDateTimePreciseValueDoShardingTest {
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithmTest.java:162]: could these test methods be split into test classes?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithmTest.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Collection;
+import java.util.Properties;
+
+import static org.junit.Assert.assertEquals;
+
+public class CosIdModShardingAlgorithmTest {
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithmTest.java:32]: class should be final

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import me.ahoo.cosid.provider.LazyIdGenerator;
+import org.apache.shardingsphere.sharding.algorithm.sharding.cosid.CosIdAlgorithm;
+import org.apache.shardingsphere.sharding.spi.KeyGenerateAlgorithm;
+
+import java.util.Properties;
+
+/**
+ * CosId key generate algorithm.
+ */
+public final class CosIdKeyGenerateAlgorithm implements KeyGenerateAlgorithm {
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java:29]: I want to know how it works in the cluster mode.

**Review by menghaoranss — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    private volatile LazyIdGenerator cosIdProvider;
+
+    /**
+     * Get properties.
+     *
+     * @return properties
+     */
+    @Override
+    public Properties getProps() {
+        return props;
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/CosIdKeyGenerateAlgorithm.java:44]: @Ahoo-Wang Maybe use @getter @setter  annotation is better.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithm.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.shardingsphere.sharding.api.sharding.standard.RangeShardingValue;
+import org.apache.shardingsphere.sharding.api.sharding.standard.StandardShardingAlgorithm;
+
+import java.util.Collection;
+import java.util.Properties;
+
+/**
+ * Modular sharding algorithm.
+ */
+public class CosIdModShardingAlgorithm<T extends Number & Comparable<T>> implements StandardShardingAlgorithm<T> {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdModShardingAlgorithm.java:29]: @Ahoo-Wang Please add final for CosIdModShardingAlgorithm.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/PropertiesUtil.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.shardingsphere.sharding.algorithm.sharding.cosid;
+
+import com.google.common.base.Preconditions;
+
+import java.util.Properties;
+
+/**
+ * Properties tool class.
+ */
+final class PropertiesUtil {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/PropertiesUtil.java:25]: @Ahoo-Wang Please add public for PropertiesUtil class and getRequiredValue method.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+     * Get type.
+     *
+     * @return type
+     */
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    static class AlgorithmConfig {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/CosIdIntervalShardingAlgorithm.java:196]: @Ahoo-Wang add private for AlgorithmConfig?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]

**Review by menghaoranss — APPROVED** [review]

**Review by TeslaCN — COMMENTED** [review]
Reminder: The license files under `shardingsphere-distribution` should be updated since new dependencies introduced.

**Comment by menghaoranss** [comment]
> Reminder: The license files under `shardingsphere-distribution` should be updated since new dependencies introduced.

@TeslaCN I will follow up.

**Comment by Ahoo-Wang** [comment]
> Reminder: The license files under `shardingsphere-distribution` should be updated since new dependencies introduced.

add to `shardingsphere-distribution/shardingsphere-jdbc-distribution/src/main/release-docs/licenses/LICENSE-cosid.txt` ?

**Pull request merged**